networkit/cpp/io/FastKONECTGraphReader.cpp

```diff
@@ -0,0 +1,270 @@
+
+			// Parse graph format directed / undirected
+			graphFormat = scanWord();
+
+			if (graphFormat == "sym" || graphFormat == "bip"){
+				directed = false;
+			}else if(graphFormat != "asym"){
+				throw std::runtime_error("Graph format not tagged properly. Format is: " + graphFormat);
+			}
+			skipWhitespace();
+			// Parse graph weighting
+			graphType = scanWord();
+			//NOTE: This parser disregards the edge weight classification of "interval scale" and "ratio scale"
+			if (graphType == "weighted" || graphType == "posweighted" || graphType == "signed"){
+				weighted = true;
+				valuesPerLine = 3;
+			} else if (graphType == "positive"){ //multiple edges
+				if(multipleEdgesHandlingMethod == SUM_WEIGHTS_UP){
+					weighted = true;
+				}
+				multiple = true; //weights will be added
+			} else if (graphType == "multisigned" || graphType == "multiweighted" || graphType == "multiposweighted"){
+				weighted = true;
+				multiple = true;
+				valuesPerLine = 3;
+			} else if (graphType == "dynamic"){
+				throw std::runtime_error("Dynamic graphs are not supported yet");
+			} else if (graphType != "unweighted"){
+				throw std::runtime_error("Graph weight not tagged properly. Weight is: " + graphType);
+			}
+			DEBUG("First property line read in. Format: "+graphFormat+" / Type: "+graphType);
+			if (graphFormat == "bip"){
+				INFO("Your graph is flagged as a bipartite one. Keep in mind that"
+				 "NetworKit cannot handle this kind of graph in its special cases."
+				 "It is imported as a usual undirected graph and edges might be discarded.");
+			}
+			if(multiple){
+				DEBUG("Selected handling method for multiple edges is: "+std::to_string(multipleEdgesHandlingMethod));
+			}
+		} else {
+			throw std::runtime_error("No graph properties line found. This line is mandatory.");
+		}
+
+		skipWhitespace();
+		if (*it != '\n'){
+			throw std::runtime_error("No break symbol after first property line");
+		}else{
+			++it;
+		}
+		skipWhitespace();
+		//second optional property line
+		if(*it == '%'){
+			++it;
+			skipWhitespace();
+			numberOfEdges = scanId();
+			skipWhitespace();
+			numberOfNodes = scanId();
+			secondPropertyLine = true;
+			while(it != end && *it != '\n')
+				++it;
+			if(it >= end){ // proper error message if file ends unexpected
+				throw std::runtime_error("Unexpected end of file");
+			}
+			if (*it != '\n'){
+				throw std::runtime_error("File not properly formatted. Last character in second property line is: "+std::string(1, *it));
+			}else{
+				++it;
+			}
+			DEBUG("Second property line read in. Edges: "+std::to_string(numberOfEdges)+ " / Nodes: "+std::to_string(numberOfNodes));
+		}
+
+
+		Graph graph((secondPropertyLine ? numberOfNodes : 0), weighted, directed);
+
+		//Map nodes and increase graph size if no second property is defined
+		auto mapNode = [&] (node in) -> node {
+			if(secondPropertyLine){
+				if (in > numberOfNodes){
+					return graph.hasNode(in) ? in : graph.addNode();
```

Comment:
Why are new nodes added dynamically here? And if you see the same node id above the range, why are always new nodes added? Is it somewhere defined (I have never dealt with KONECT graphs) that these nodes outside the range always occur only once?

Reply:
Fixed in https://github.com/kit-parco/networkit/pull/169/commits/2af7dbc977feb8f954a6706ba93c009295ecf4e0

I agree, that the behaviour there was misleading and could lead to misbehaviour in these special cases where the file header are corrupted. Now, I implemented a "fallback" as if no header was given in first place